### PR TITLE
VIZ-3722 trajectory functions

### PR DESCRIPTION
**Pull request by aswilkinson**
Added functions to get the distance, angle, and dihedral angle between atoms over chosen frames.

Also added tests for testing said functions.

Note: "get_(insert)_string" functions are currently deprecated, and need to be updated. The current ones are patterned the same as the original "get_frame_string"

### CODE REVIEW

File: mdsrv/trajectory.py [hunk 1]
```diff
@@ -188,6 +189,93 @@ def get_path_string( self, atom_index, frame_indices=None ):
...
+            frame_indices.sort()
+            size = len( frame_indices )
+            frames = map( int, frame_indices )
+        else:
+            size = self.numframes
+            frames = range( size )
+        dist = np.zeros(size, dtype=np.float32)
+        counter = 0
+        for i in frames:
+            box, coords, time = self._get_frame( i )
```
Comment (papillot) [inline mdsrv/trajectory.py:203]: Note that this method has an optional argument: atomIndices. You can pass an array containing atom_index0 and atom_index1 to receive an nparray with the coordinates from these points.

Note that the syntax of the argument is a bit special. It is optimized to pass ranges! For example, if you pass the following list `[1,5,9,10]` it will give you the atoms `[1,2,3,4,5,9,10]`. This means that if you want the indices 12 and 54 only, you'll have to provide `[12,12,54,54]`

File: mdsrv/trajectory.py [hunk 2]
```diff
@@ -188,6 +189,93 @@ def get_path_string( self, atom_index, frame_indices=None ):
...
+            for j in range(3):
+                dist[counter] += (coords[ atom_index0 ][ j ] - coords[ atom_index1 ][ j ])**2
+            dist[counter] = math.sqrt(dist[counter])
+            counter += 1
+        return dist
+
+    def get_distance_string( self, atom_index0, atom_index1, frame_indices=None ):
+        dist = self.get_distance( atom_index0, atom_index1, frame_indices=frame_indices )
```
Comment (papillot) [inline mdsrv/trajectory.py:211]: `frame_indices` as a named argument can be omitted here (does not help readability)

File: mdsrv/trajectory.py [hunk 3]
```diff
@@ -188,6 +189,93 @@ def get_path_string( self, atom_index, frame_indices=None ):
...
+            size = len( frame_indices )
+            frames = map( int, frame_indices )
+        else:
+            size = self.numframes
+            frames = range( size )
+        ang_calc = np.zeros(size, dtype=np.float32 )
+        counter = 0
+        for i in frames:
+            box, coords, time = self._get_frame( i )
+            vector0 = np.array(coords[atom_index0]) - np.array(coords[atom_index1])
```
Comment (papillot) [inline mdsrv/trajectory.py:226]: I think that coords is already an np.array

File: mdsrv/trajectory.py [hunk 4]
```diff
@@ -188,6 +189,93 @@ def get_path_string( self, atom_index, frame_indices=None ):
...
+            vector2 = np.array(coords[atom_index2]) - np.array(coords[atom_index1])
+            dot_product = vector0[0] * vector2[0] + vector0[1] * vector2[1] + vector0[2] * vector2[2]
+            magnitude0 = math.sqrt(vector0[0]**2 + vector0[1]**2 + vector0[2]**2)
+            magnitude1 = math.sqrt(vector2[0]**2 + vector2[1]**2 + vector2[2]**2)
```
Comment (papillot) [inline mdsrv/trajectory.py:230]: use numpy routines here `np.sqrt(np.sum(np.square(vector0)))` https://numpy.org/doc/stable/reference/routines.math.html#mathematical-functions

File: mdsrv/trajectory.py [hunk 5]
```diff
@@ -188,6 +189,93 @@ def get_path_string( self, atom_index, frame_indices=None ):
...
+
+            cosine_angle = dot_product / (magnitude0 * magnitude1)
+            angle = math.acos(cosine_angle)
+            degrees = math.degrees(angle)
+            ang_calc[counter] = degrees
```
Comment (papillot) [inline mdsrv/trajectory.py:235]: These can be done outside of the loop, by collecting the results in a vector and by using numpy to treat all data points at the same time.

File: mdsrv/trajectory.py [hunk 6]
```diff
@@ -188,6 +189,93 @@ def get_path_string( self, atom_index, frame_indices=None ):
...
+            counter += 1
+        return ang_calc
+    
+    def get_angle_string( self, atom_index0, atom_index1, atom_index2, frame_indices=None ):
+        ang = self.get_angle( atom_index0, atom_index1, atom_index2, frame_indices=frame_indices )
```
Comment (papillot) [inline mdsrv/trajectory.py:240]: same suggestion as above, you can remove the repetition for the named argument

File: mdsrv/trajectory.py [hunk 7]
```diff
@@ -188,6 +189,93 @@ def get_path_string( self, atom_index, frame_indices=None ):
...
+        else:
+            size = self.numframes
+            frames = range( size )
+        dihed_calc = np.zeros(size, dtype=np.float32 )
+        counter = 0
+        for i in frames:
+            box, coords, time = self._get_frame( i )
+            vector0 = np.array(coords[atom_index1]) - np.array(coords[atom_index0])
+            vector1 = np.array(coords[atom_index2]) - np.array(coords[atom_index1])
+            vector2 = np.array(coords[atom_index3]) - np.array(coords[atom_index2])
```
Comment (papillot) [inline mdsrv/trajectory.py:257]: coords are already np arrays

File: mdsrv/trajectory.py [hunk 8]
```diff
@@ -188,6 +189,93 @@ def get_path_string( self, atom_index, frame_indices=None ):
...
+            frames = map( int, frame_indices )
+        else:
+            size = self.numframes
+            frames = range( size )
+        dist = np.zeros(size, dtype=np.float32)
+        counter = 0
+        for i in frames:
+            box, coords, time = self._get_frame( i )
+            for j in range(3):
+                dist[counter] += (coords[ atom_index0 ][ j ] - coords[ atom_index1 ][ j ])**2
```
[the remark this answers is not in the document]
Comment (papillot) [reply mdsrv/trajectory.py:205]: in fact the linalg.norm routine will work here

File: mdsrv/trajectory.py [hunk 9]
```diff
@@ -188,6 +189,93 @@ def get_path_string( self, atom_index, frame_indices=None ):
...
+            
+            # Normalize the normal vectors
+            normal0 = normal0 / np.linalg.norm(normal0)
+            normal1 = normal1 / np.linalg.norm(normal1)
+            
+            dot_product = np.dot(normal0, normal1)
+            angle = np.arccos(dot_product)
+            triple_product = np.dot(np.cross(normal0, normal1), vector1)
+            if triple_product < 0:
+                angle = -angle
```
Comment (papillot) [inline mdsrv/trajectory.py:269]: Nice! You've found it!

File: tests/test_trajectory.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    atom0, atom1, frame = 93, 74, [0]
+    assert str(round(TESTING_FILE.get_distance(atom0, atom1, frame)[0], 2)) == '3.83'
+
+def test_angle():
+    atom0, atom1, atom2, frame = 23, 74, 138, [10]
+    assert str(round(TESTING_FILE.get_angle(atom0, atom1, atom2, frame)[0], 1)) == '143.4'
+
+def test_dihedral():
+     atom0, atom1, atom2, atom3, frame = 74, 23, 127, 145, [15]
+     assert str(round((TESTING_FILE.get_dihedral(atom0, atom1, atom2, atom3, frame))[0], 1)) == '23.8'
```
Comment (papillot) [inline tests/test_trajectory.py:24]: Very nice!
Please consider also testing for the negative dihedral by changing the order of the atoms from this same test.

**Review by papillot — CHANGES_REQUESTED** [review]
That's a great work! An excellent stepping stone for this task of adding measures to MDsrv. The code you wrote is easily readable, with helpful comments when need be!

Here are some requests and advices:
- request: keep the original webapp.html file, and rename "ours" to example.html
- advice: make distinct commits for distinct parts of your work. You can have for example on commit named "typo" for the fix you did to the `get_struc_directoy` function (very nice to have done that by the way). Another for the .gitignore, etc... This would help if you ever need to undo one of those commits (for example by doing a revert or by using `git rebase --interactive`
- request: for performance reasons, try to write your code using numpy as much as possible. We can hope gains of one order of magnitude. In some cases that means you can group the coordinates collection phase to store values in your vectors, and then work on the vectors for the full trajectory directly. The best approach before doing these last optimizations is to benchmark your code first, to get an idea of the time it takes to process a trajectory. So, for now, let's try to get rid of the inner for loops
- request: the first 8 lines of these functions are identical. they should be abstracted in a function that returns 2 objects: the frames and the (empty) measures

**Review by papillot — CHANGES_REQUESTED** [review]
Please remove the last 3 commits and review the way the routing mechanism is done, to keep the <root> feature.
Review also the format for the `frameindices` variable. 

For later, the logic to get the path and the frame indices should be isolated in a common place (DRY). Maybe a function that would return the `TRAJ_CACHE.get(path, struc_path)` instance and the frame_indices?

File: tests/test_trajectory.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python
+import sys
+from pathlib import Path
+path_root = Path(__file__).parents[1]
+sys.path.append(str(path_root) + "/mdsrv")
+from trajectory import *
+import pytest
+import json
+import numpy as np
```
[the remark this answers is not in the document]
Comment (aswilkinson) [reply tests/test_trajectory.py:9]: I imported json and numpy originally for testing outputs, but the first one (line six) is needed for the imports to work properly (even if VScode doesn't properly recognize it).
